contrib/HDRMerge.lua

```diff
@@ -0,0 +1,348 @@
+    label = 'Preview Size', 
+    tooltip = 'Output file\'s built-in preview size \ndefault: '..pref_size,
+    value = pref_size_enum,
+    "half", "full", "none",        
+	}  
+HDRMerge_lbl_batch= dt.new_widget("section_label"){
+     label = "Batch Options",
+	}
+HDRMerge_chkbtn_batch = dt.new_widget("check_button"){
+    label = '  Batch Mode', 
+    value = false, --pref_batch,
+    tooltip ='Operate in batch mode. When operating in batch mode output files will NOT be auto-imported \ndefault: (false).',  
+	}
+HDRMerge_sldr_gap = dt.new_widget("slider"){
+	label = 'Batch Gap [sec.]',
+	tooltip = 'Gap, in seconds, between batch mode groups \ndefault: (3).',
+	soft_min = 1,
+	soft_max = 60,
+	hard_min = 1,
+	hard_max = 60,
+	step = 1,
+	digits = 0,
+	value = 3, --pref_gap,
+	}
+HDRMerge_lbl_out= dt.new_widget("section_label"){
+     label = "Auto-import Options",
+	}
+HDRMerge_cmbx_style = dt.new_widget("combobox"){
+	label = "Apply Style on Import",
+	tooltip = "Apply selected style on auto-import to newly created HDRMerge DNG\nDoes not apply when in Batch mode.",
+	value = 1,
+	"none",
+	}
+pref_style_enum = 1
+for k=1, (styles_count-1) do
+	HDRMerge_cmbx_style[k+1] = styles[k].name
+	if styles[k].name == pref_style then pref_style_enum = k+1 end
+end
+HDRMerge_cmbx_style.value = pref_style_enum
+HDRMerge_chkbtn_cpytags = dt.new_widget("check_button"){
+    label = '  Copy Tags', 
+    value = dt.preferences.read("module_HDRMerge", "copy_tags", "bool"),
+    tooltip ='Copy tags from first image. When operating in batch mode this will NOT be performed. \ndefault: (true).',  
+	}
+HDRMerge_lbl_tags= dt.new_widget("label"){
+     label = 'Additional Tags',
+	}
+HDRMerge_entry_tag = dt.new_widget("entry"){
+	tooltip = "Additional tags to be added on import. Seperate with commas, all spaces will be removed",
+	text = dt.preferences.read("module_HDRMerge", "add_tags", "string"),
+	placeholder = "Enter tags, seperated by commas",
+	editable = true
+}
+HDRMerge_btn_run = dt.new_widget("button"){
+	label = "Run HDRMerge",
+	tooltip = "Runs HDRMerge on selected images, with selected settings",
+	clicked_callback = function() HDRMerge() end
+	}
+HDRMerge_lbl_note= dt.new_widget("label"){
+     label = 'Defaults can be adjusted under:\n"Settings > lua options"',
+	}
+dt.register_lib(
+	"HDRMerge_Lib",	-- Module name
+	"HDRMerge",	-- name
+	true,	-- expandable
+	false,	-- resetable
+	{[dt.gui.views.lighttable] = {"DT_UI_CONTAINER_PANEL_RIGHT_CENTER", 99}},	-- containers
+	dt.new_widget("box"){
+		orientation = "vertical",
+		HDRMerge_lbl_base,
+		HDRMerge_cmbx_bps,
+		HDRMerge_cmbx_size,
+		HDRMerge_lbl_batch,
+		HDRMerge_chkbtn_batch,
+		HDRMerge_sldr_gap,
+		HDRMerge_lbl_out,
+		HDRMerge_cmbx_style,
+		HDRMerge_chkbtn_cpytags,
+		HDRMerge_lbl_tags,
+		HDRMerge_entry_tag,
+		HDRMerge_btn_run,
+		HDRMerge_lbl_note
+	}
+)
+
+-- PREFERENCES --
+dt.preferences.register("module_HDRMerge", "add_tags",	-- name

```

Comment:
You don't need to register the preferences.  You can just use a callback on your gui elements and dt.preferences.write() to save them.  This gets rid of the necessity to restart the program just to use the module.

Reply:
I'll make this change.

Reply:
I refactored this script and the new version only registers a preference for the binary location. However, I was thinking that even the binary location may not need to be registered on all OS types. I know it is needed on Windows, I believe it is NOT needed on linux, I have no clue about MacOS. If we make that determination I can wrap the binary preference registry in an if statement based on OS type.

Reply:
I think the only time you should use dt.register_preference() is if you don't have a user interface, i.e. it's only a shortcut, or a button on the selected images module. If it's an exporter or a module, then put the preferences in the UI. The problem with dt.register_preference() is that you have to exit darktable in order for the preference to take effect.  If you have a user interface, then put the executable path widget in the UI and save the preference right away so the use can use the script without having to restart darktable.  That's why I created df.executable_path_widget(), so all that could be done and included in the UI with a minimum of coding.  The only thing that would make it better is if you ran the script without the executable path and it threw an error and drew a box around the executable_path_preference in red so that the user knew exactly where to go to fix it.

Reply:
Ok, I can add an executable path widget to the module rather than in preferences. I just though it was cleaner long-term to have the executable in the preferences. You only have to set it once and then never again, so it seemed to me having it displayed in the UI of the widget for just a single use was extraneous. And you already have to go in to the Luarc file to install new widgets, so I figured one more "install step" was a minor price to pay for the cleaner UI. 

There may be a best of both worlds, I'll have to play. I notice preferences.register() has a "set_callback" I wonder if this can be used to trigger updating something that contains the executable path.

Reply:
I don't get the luarc/new widget thing.  Did you mean scripts?  If so, copy tools/script_manager.lua to your luarc, in effect making it a replacement.  Now you never have to edit your luarc again.  You just enable and disable scripts from the script_manager interface.  As a bonus for developers, if you have a syntax error and your script fails to load, you can fix the syntax error, then click enable <script> again and it will try and load it again without having to stop and start darktable.  If you have a logic error, then you have to stop and start darktable because I still haven't figured out how to unload a script (actually unload a registered anything because there aren't functions to do that, until I figure you how to add unregister :-D)

Part of my thought process is big picture.  Let's say  we have 50 scripts using on average 1.5 executables which would work out to 75 preferences.  Unfortunately of those 75, 30 of them are multiple definitions of the same executable.  For instance align_image_stack is used by contrib/image_stack.lua, official/enfuse.lua, contrib/enfuseAdvanced.lua, and maybe more.  exiftool is also used multiple times.  So, the user has to figure out which one to change because they don't realize they are really the same.  Programmers have to put them in there because they can't rely on the users knowing the one exiftool executable preference is as good as another. And, the user has to go away from what he was doing and dig through the configuration data to find where to make the change, then make the change.  And then he has to restart darktable, which is my biggest hangup about register preferences.  My belief is that the user should start darktable once and use it until they're done and not have to restart it when they want to use a script, which is the reason I wrote script manager.  User's shouldn't have to know lua in order to enable the scripts.

The whole gui element can really be solved with a call to check_if_bin_exists().  If it does, don't load the executable_path_widget.  So therefore I could write a new script maybe called compare :-) and use exiftool and if it's already defined by another script then I don't ever need to ask the user for it, but the code's there just in case.  

If you decide to do it this way, then to test it you'll have to edit your darktablerc file and remove the "lua/executable_paths/<script>=" lines, or just remove the values.

Reply:
I don't use the script manager; I wasn't really aware of it or what it does. Sounds useful though. I'll have to load it up. But, as I don't use it right now I go in an manually modify my LUARC file for every new script. Your point on the big picture is a very valid one. I admit I still don't thing a single restart once you install a new script is a big deal, however, I certainly agree that having an over abundance of preferences, and likely a bunch of dupilcates, is bad, so I'm on board with changing the way executables are defined, especially if it can be done in a way that is minimally invasive to the normal module GUI. My past scripts have certainly been culprits of over-utilizing preferences and that was part of the driving reason I wanted to refactor HDRMerge and enfuseAdvanced. 

I had previously thought about wrapping the preferences.register() in an if statement that checks if it already exists. The only problem I thought about with that was the following: what if the user accidentally (or intentionally) assigns a different executable, and then the ability to correct it disappears? For example: On my dev laptop I didn't have HDRMerge installed for a long time, but I still tested my HDRMerge script by just pointing to hugin.exe. Maybe this is a corner case we don't need to care about...

If we either don't care about that, or find a way around it then I could imagine doing the following:
putting all gui elements in box containers inside a stack. One box holds your normal GUI elements, the other just executable choosers. When the script loads it checks if the bin exists. If it doesn't then the executable chooser container is made active, otherwise the normal GUI is active. Once the user sets the executable path and it's verified with a check_if_bin_exists() the GUI container is made active. Thus the path choosers are always hidden by being not active in the stack unless we detect they need to be shown at startup.

Reply:
You really should load script_manager.  It lets you see all that is available, and if you hover over the button it tells you what it does.  Then you could look at tools/executable_manager.lua which provides a way to manipulate all the executable path preferences and redefine the location or clear it if necessary.

The nice thing about script manager is it rescans the lua directory every time you start darktable, so it picks up new scripts automatically.  You can also add extra directories like development, testing, mynewtoys, etc. and it picks those up too and just adds new categories.  The best thing about script_manager, in my opinion, is that a misbehaving script doesn't stop other scripts from loading.  If you use the luarc file and your first script has a syntax error, then nothing loads.  

I remember when I first tried to use the scripts and I didn't know lua.  It took 20 or 30 minutes before I finally figured it out, starting with # isn't a comment :-).  At that point I realized that the lua-scripts weren't really accessible to normal users, unless you had a programming background or a lot of perseverance.  I wrote the first script_manager a few months later and have been using that as my lua-script interface ever since.  It went through several iterations before I thought it was fit for public consumption, but I finally got there.

The other feature I use a lot, is that I only turn on what I'm using and when I'm done I turn it off. If I'm in the middle of an editing session and I decide I want to finish an image in gimp, I just to go script_manager, turn it on, then use it, all without having to restart darktable and lose my train of thought. :-D

Reply:
I just loaded it up and had a quick play with it. That is very nice, very impressive! I'll start using this. I'll have to setup GIT on my main computer for that.

Let me know if you have any thoughts on how to avoid the situation described above where a user might accidentally select the wrong BIN and then the availability to change it disappears. If you think that's rare enough of a case that we can ignore it and be fine, I'll go ahead and start working on these changes across multiple of my scripts.

Reply:
If a user accidentally chooses the wrong executable, they can use tools/executable_manager to fix it.  executable_manager lets you change any executable that has been used in the lua scripts.  It has a combobox for all of the executables it knows about.  Select the executable and an executable path widget is displayed along with a clear button.  Select the correct executable, or clear the entry.

Reply:
Ahh perfect! Well then I'll start working on a different way of prompting user for executable paths in my scripts.